### ISSUE TITLE
500 Error when opening websocket connection

### ISSUE THREAD

**Issue by ysica** [eclipse/jetty.project#3827]
I'm trying to open a websocket connection to my Jetty app (websocket echo), but I get an HTTP 500 error when I try to open the connection. Here's the associated exception that's thrown on the server:

```
java.lang.ClassCastException: org.eclipse.jetty.util.DecoratedObjectFactory cannot be cast to org.eclipse.jetty.util.DecoratedObjectFactory
	at org.eclipse.jetty.websocket.server.WebSocketServerFactory.doStart(WebSocketServerFactory.java:308)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.websocket.servlet.WebSocketServlet.init(WebSocketServlet.java:142)
	at javax.servlet.GenericServlet.init(GenericServlet.java:244)
	at org.eclipse.jetty.servlet.ServletHolder.initServlet(ServletHolder.java:665)
	at org.eclipse.jetty.servlet.ServletHolder.getServlet(ServletHolder.java:514)
	at org.eclipse.jetty.servlet.ServletHolder.prepare(ServletHolder.java:798)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:530)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:190)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:219)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:126)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:530)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:347)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:256)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:382)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
	at java.lang.Thread.run(Thread.java:745)
```

Here are some things worth noting:
- I'm using Jetty 9.4.8v20171121 - I've tried 9.4.19.v20190610 as well, but the behavior is the same
- This error is thrown only when Jetty is running directly on a Windows server - I can successfully use this websocket echo when running the project on a Linux server, as well as through the Jetty Maven plugin on Windows
- I tried removing jetty jar files that didn't match my Jetty version under webapp/WEB-INF with no luck (there were some jetty-6.1.26 jar files in that directory)

Does anyone have any ideas of how I should go about diagnosing this problem so that I can get this websocket echo project working on Windows?

**Comment by joakime** [comment]
Sounds like you have jetty server jars in your `WEB-INF/lib`.
Remove them.

**Comment by ysica** [comment]
Thanks for the help joakime. I had these jetty dependencies in WEB-INF/lib:

- jetty-6.1.26.jar
- jetty-client-9.4.8v20171121.jar
- jetty-http-9.4.8v20171121.jar
- jetty-io-9.4.8v20171121.jar
- jetty-security-9.4.8v20171121.jar
- jetty-server-9.4.8v20171121.jar
- jetty-servlet-9.4.8v20171121.jar
- jetty-util-6.1.26.jar
- jetty-util-9.4.8v20171121.jar
- jetty-webapp-9.4.8v20171121.jar
-jetty-xml-9.4.8v20171121.jar

I was able to remove all but jetty-util-9.4.8v20171121.jar, jetty-io-9.4.8v20171121.jar and jetty-http-9.4.8v20171121.jar (code in the project depended on these libraries). After removing all the jetty dependencies I could, I'm still running into the same error unfortunately.

**Comment by joakime** [comment]
None of those should be in your `WEB-INF/lib`
That is the source of your problem, the conflicts on classes between your webapp and the server.

**Comment by ysica** [comment]
I gave a quick look at removing those three jar files and the app doesn't start without them. It's a large app, so I'll have to take a more detailed look at removing those dependencies.

I found the output of this command a little puzzling:

```
> java -jar start.jar --list-config
-snip-
Jetty Server Classpath:
-----------------------
Version Information on 35 entries in the classpath.
Note: order presented here is how they would appear on the classpath.
      changes to the --module=name command line options will be reflected here.
 0:      1.4.1.v201005082020 | ${jetty.base}\lib\mail\javax.mail.glassfish-1.4.1.v201005082020.jar
 1:                    (dir) | ${jetty.base}\resources
 2:                    3.1.0 | ${jetty.base}\lib\servlet-api-3.1.jar
 3:                 3.1.0.M0 | ${jetty.base}\lib\jetty-schemas-3.1.jar
 4:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-http-9.4.8.v20171121.jar
 5:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-server-9.4.8.v20171121.jar
 6:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-xml-9.4.8.v20171121.jar
 7:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-util-9.4.8.v20171121.jar
 8:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-io-9.4.8.v20171121.jar
 9:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-jndi-9.4.8.v20171121.jar
10:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-security-9.4.8.v20171121.jar
11:                      1.2 | ${jetty.base}\lib\transactions\javax.transaction-api-1.2.jar
12:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-servlet-9.4.8.v20171121.jar
13:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-webapp-9.4.8.v20171121.jar
14:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-plus-9.4.8.v20171121.jar
15:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-annotations-9.4.8.v20171121.jar
16:                      6.0 | ${jetty.base}\lib\annotations\asm-6.0.jar
17:                      6.0 | ${jetty.base}\lib\annotations\asm-commons-6.0.jar
18:                      1.2 | ${jetty.base}\lib\annotations\javax.annotation-api-1.2.jar
19:    3.12.3.v20170228-1205 | ${jetty.base}\lib\apache-jsp\org.eclipse.jdt.ecj-3.12.3.jar
20:          9.4.8.v20171121 | ${jetty.base}\lib\apache-jsp\org.eclipse.jetty.apache-jsp-9.4.8.v20171121.jar
21:                   8.5.20 | ${jetty.base}\lib\apache-jsp\org.mortbay.jasper.apache-el-8.5.23.jar
22:                      2.3 | ${jetty.base}\lib\apache-jsp\org.mortbay.jasper.apache-jsp-8.5.23.jar
23:                    1.2.5 | ${jetty.base}\lib\apache-jstl\org.apache.taglibs.taglibs-standard-impl-1.2.5.jar
24:                    1.2.5 | ${jetty.base}\lib\apache-jstl\org.apache.taglibs.taglibs-standard-spec-1.2.5.jar
25:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-client-9.4.8.v20171121.jar
26:          9.4.8.v20171121 | ${jetty.base}\lib\jetty-deploy-9.4.8.v20171121.jar
27:                      1.0 | ${jetty.base}\lib\websocket\javax.websocket-api-1.0.jar
28:          9.4.8.v20171121 | ${jetty.base}\lib\websocket\javax-websocket-client-impl-9.4.8.v20171121.jar
29:          9.4.8.v20171121 | ${jetty.base}\lib\websocket\javax-websocket-server-impl-9.4.8.v20171121.jar
30:          9.4.8.v20171121 | ${jetty.base}\lib\websocket\websocket-api-9.4.8.v20171121.jar
31:          9.4.8.v20171121 | ${jetty.base}\lib\websocket\websocket-client-9.4.8.v20171121.jar
32:          9.4.8.v20171121 | ${jetty.base}\lib\websocket\websocket-common-9.4.8.v20171121.jar
33:          9.4.8.v20171121 | ${jetty.base}\lib\websocket\websocket-server-9.4.8.v20171121.jar
34:          9.4.8.v20171121 | ${jetty.base}\lib\websocket\websocket-servlet-9.4.8.v20171121.jar
-snip-
```

It seems like the app should have no problem starting when I remove the three Jetty jar files from WEB-INF/lib because they're in the Jetty classpath. Do you know why I'd be running into issues with this?

**Comment by gregw** [comment]
@ysica just because a jar is on the jetty classpath does not mean it is exposed to the webapp.

You probably need to remove those 3 jars and then use then exclude exactly the classes you need to see from the serverClasses configuration.  This is normally done for you with the websocket module.

**Comment by ysica** [comment]
I figured out the cause of the problem. I have a Jetty-level start.ini config file under opt/jetty which includes the websocket module (`--module=websocket`). There's also a start.ini file under opt/webappbase which didn't include the websocket module. I added `--module=websocket` to opt/webappbase/start.ini and restarted the server, and then my websocket echo test began to work.

🤦‍♂️

Thanks for the help guys!
